json-format.md

```diff
@@ -141,22 +141,55 @@ For any other type, the implementation MUST translate the data value into a
 [JSON value][json-value], and use the member name `data` to store it inside the
 JSON object.
 
+If the implementation determines that the type of data is NOT `Binary`, it MUST
+next determine whether the `datacontenttype` attribute, if present, indicates a
+JSON-oriented content type. A JSON-oriented content type is one whose subtype
+is `json` or has a JSON format; that is, a MIME type of the form `*/json` or
+`*/*+json`. If the `datacontenttype` attribute is absent, it is treated as if it
+were set to `application/json`, which is a JSON-oriented content type.
+
+If the implementation determines that the `datacontenttype` of an event with
+non-`Binary` data is JSON-oriented, it MUST translate the data value directly to
+a [JSON value][json-value], and use the member name `data` to store it inside
+the JSON object. It MUST store the JSON value directly; it MUST NOT serialize it
```

Comment:
Effectively, yes. It's tempting to serialize the json object and store it as a string, but that is incorrect and changes the value. Do you have any suggestions on the wording?

Reply:
I found this a bit hard to parse too. What if we said something like:
```
...
non-Binary data is JSON-oriented, it MUST parse the data value as a
JSON value. Note that the parsing MUST only use the characters
provided, it MUST NOT assume the intent of the data by, for example,
adding quotes around a string. If the provided data can not be parsed
as valid JSON then it MUST be treated as an invalid Cloud Event.
```

Reply:
@duglin That assumes the JSON-valued data is expressed in memory (or was decoded from an event source) as a serialized string (i.e. that would need to be parsed in order to encode into JSON format.) I don't think we should make that assumption. Similarly, the clause about just passing the data value through if the event is "simply being forwarded" also assumes that the event was originally sourced from JSON format, which might not be true.

I was trying to word this in such a way as to start with a "decoder-agnostic" event representation. I assume (although maybe I'm wrong?) that most SDKs have some kind of event "class" or similar construct that represents an event in memory, independent of where the event originally came from (whether it was originally decoded from some format, or maybe was just constructed directly by application code). Given such an in-memory construct and however it might represent the event "data" in memory, how do we say that when serializing to JSON format, the `data` attribute should be set to an actual JSON value rather than a serialized string?

Reply:
I was thinking more of it being an array of bytes and in that case there are two options:
1 - pass it along "as is" in which case whether it's valid JSON or not doesn't matter to the current component (the "Note" section I added)
2 - it's parsed by a JSON parser. In which case it needs to only use the exact bytes provided, it can't (as I said) add quotes to turn it into `hello world` into a JSON string.

But now that I'm rereading the section... are we talking about just a structured CE or unstructured? I think I was assuming unstructured in which case treating it as an array of bytes seems correct.  If we're talking structured, then I guess we don't need to say anything since it's all just normal JSON.

I'm also now wondering if this section is talking about serializing or deserializing :-)   Your text (I think) is talking about deserializing, but the paragraph right before is serializing.  Maybe we need to be clear about this too - maybe even split it.

Reply:
This section is still talking about serialization. Okay it's clear I need to try writing this paragraph again. Maybe it would also be a good idea to add subsection headers for serialization and deserialization to further clarify.

Reply:
Pushed a new commit: I added subsection headers for serialization and deserialization to try to reduce the confusion, clarified some of the text with an example, and added a note on failure cases.